sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala

```diff
@@ -112,15 +114,26 @@ object ParquetUtils {
       metadata: Seq[FileStatus],
       commonMetadata: Seq[FileStatus])
 
-  private def splitFiles(allFiles: Seq[FileStatus]): FileTypes = {
-    val leaves = allFiles.toArray.sortBy(_.getPath.toString)
```

Comment:
It was actually @liancheng's advice (wow 5 years ago). Sorting it will gives you a deterministic result. Relying on the native listing order doesn't look a great idea. From a cursory look, HDFS's vs Java's `File.list` (by `RawLocalFileSystem `) behaviours look inconsistent as an example.

In addition, sorting itself won't cause a lot of performance penalty given most of performance penalty comes from listing up files, and given that the number of files should be, really, really high which isn't a pretty common case.

Reply:
Also, IIRC, `sortBy` uses Timsort under the hood which gives you linear time complexity when the input is already sorted, which I believe are in most cases here.

Reply:
IIUC, files's deterministic order only matters when we need to merge schema? If so, I think this PR doesn't introduce any behavior change on files' order. We'll still get a deterministic order when we need it.

Reply:
And thanks for providing performance penalty knowledge around here. I agree that if there' no obviously performance improvement, this may not be such necessary change.

Reply:
Yes, that issue was specific to the ordering; however, it can cause a behaviour change: we will pick one file to read the schema, right?

It is possible to put the complete parquet file in the top in a directory, and other files are partial. For instance,

```scala
scala> spark.read.parquet("test/a.snappy.parquet")
res1: org.apache.spark.sql.DataFrame = [a: bigint, b: bigint]

scala> spark.read.parquet("test/b.snappy.parquet")
res2: org.apache.spark.sql.DataFrame = [b: bigint]

scala> spark.read.parquet("test").show()
+----+---+
|   a|  b|
+----+---+
|   0|  0|
|   1|  1|
|   2|  2|
|null|  0|
|null|  1|
|null|  2|
+----+---+
```

If the order is switched:

```scala
scala> spark.read.parquet("test").show()
+---+
|  b|
+---+
|  0|
|  1|
|  2|
|  0|
|  1|
|  2|
+---+
```

Reply:
Thanks for providing a specific case. But I think that user should really need to use `mergeSchema=true` option for such case if there's no guarantee that the completed parquet file should always be on top of a directory.

Besides, I believe that there's assuming that schemas should be consistent if `mergeSchema=false`. And user should be responsible for the result schema if not.

>  In either case, we fall back to any of the first part-file, and just assume all schemas are consistent.

Reply:
That's true but users always find awesome ways to use. True, this case `mergeSchema` have better to be set `true` but many people want to avoid in production because `mergeSchema` is super expensive as it needs to touch every file.

In non performance sensitive codes, we better keep the codes simple and more readable anyway.